readme.md
# MODX Vapor

This is MODX Vapor, a PHP script for Extracting a complete snapshot of a MODX site, including it's file and database artifacts for use in importing the site into MODX Cloud.


## Environment Requirements

- PHP 5.2+
- PHP Zip extension
- All requirements for MODX 2.1+


## Installation

- Install the project as an immediate subdirectory of the `MODX_BASE_PATH` for the MODX site you want to extract as a vapor package. This subdirectory is referred to as the VAPOR_DIR for the remainder of this document.


## Usage

MODX Vapor can be run from the CLI or from a web browser. It is recommended that it be run from CLI if possible however, to avoid potential issues with web server and/or PHP-related timeouts. This is especially true for larger MODX sites. In either case, the snapshot will be created in your `MODX_CORE_PATH . 'packages/'` directory with the snapshot name indicated in the output from the script.

### Running via CLI

To create a MODX Cloud-compatible snapshot of your MODX site via CLI, change directory to your `MODX_BASE_PATH` and run `VAPOR_DIR . 'vapor.php'`, e.g.

    php vapor/vapor.php

_NOTE: The `vapor/` subdirectory is the `VAPOR_DIR`_

### Running via Browser

To create a MODX Cloud-compatible snapshot of your MODX site via your browser, navigate to `MODX_SITE_URL . VAPOR_DIR . 'vapor.php'` in your browser and wait for the process to complete, e.g.

    http://localhost/revo/vapor/vapor.php

### Custom Options

There are a few options available for customizing the execution of Vapor. You can define these options by creating a `VAPOR_DIR . 'config.php'` file with the following contents:

```php
<?php
return array(
    'excludeFiles' => array(),
    'excludeExtraTables' => array(),
    'excludeExtraTablePrefix' => array(),
);
```

- __excludeFiles__ &mdash; An array of file/directory names (do not include the trailing / on directory names) to exclude from the `MODX_BASE_PATH`. Normally Vapor would package up any non-core files located within your `MODX_BASE_PATH`. Define specific items you want to skip here.
- __excludeExtraTables__ &mdash; An array of non-core tables to exclude from your database. Normally Vapor will package any non-core tables in the database.
- __excludeExtraTablePrefix__ &mdash; An array of non-core tables to not prepend with your MODX `table_prefix`. Normally, Vapor will keep track of which non-core tables need to have a table_prefix prepended when Injected into the target MODX site, but if the source MODX site does not define a `table_prefix`, Vapor has to assume all of them will need the target's `table_prefix`. Define specific tables that should not get the target's `table_prefix` upon Injection here if your source database does not use a `table_prefix`. _This is not necessary if the source does define a `table_prefix`._


## Troubleshooting

### Review the Vapor log

Each time you run vapor it will log information about it's execution into a file with the same timestamp as the zip file that gets created (i.e. the server time when vapor is run). This file will be in the format `vapor-{timestamp}.log` and located in your logs directory at `{core_path}cache/logs/` (or `{core_path}cache/{MODX_CONFIG_KEY}/logs` if using a custom MODX_CONFIG_KEY value).


## Copyright

MODX Vapor is Copyright 2012-2013 by MODX, LLC.


changelog.md
# MODX Vapor Changelog

## 1.1.1-pl (March 24, 2016)

- Additional error handling

## 1.1.0-dev

- Fix package signatures
- Additional information output to Vapor log

## 1.0.0-pl (October 30, 2012)

- Fix missing table truncation script when no extension_packages setting exists
- [#8975](http://tracker.modx.com/issues/8975) Fix case in SQL for selecting non-core tables

## 1.0.0-beta-5 (October 9, 2012)

- [#8871](http://tracker.modx.com/issues/8871) Fix drop table statement in vaporVehicles to respect table_prefix properly

## 1.0.0-beta-4 (October 8, 2012)

- [#8817](http://tracker.modx.com/issues/8817) Fix media source import via resolvers
- Fix invalid reference from $options to $fileMeta in resolvers
- [#8871](http://tracker.modx.com/issues/8871) Fix import of extra tables without a table_prefix
- Add excludeExtraTablePrefix to avoid prepending table_prefix for specific tables
- Add excludeExtraTables option for excluding specific extra non-core tables
- Add excludeFiles option for excluding specific files/dirs from base_path
- Add ability to provide vaporOptions from config file

## 1.0.0-beta-3 (September 26, 2012)

- Fix PHP warning when table_prefix is empty
- Fix realpath missing trailing slash on extension_packages

## 1.0.0-beta-2 (August 23, 2012)

- Use realpath for extension_packages path only if not absolute
- Log execution to file for easier analysis
- Use set_time_limit(0) repetitively to avoid max_execution_time errors

## 1.0.0-beta-1 (July 19, 2012)

- Initial release of MODX Vapor.

scripts/resolve.vapor_model.php
<?php
/*
 * Copyright 2012 by MODX, LLC.
 *
 * This file is part of MODX Vapor.
 *
 * Vapor is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Vapor is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Vapor; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

/**
 * @var xPDOTransport $transport
 * @var array $object
 * @var array $options
 */
$transport->xpdo->loadClass('vapor.vaporVehicle', MODX_CORE_PATH . 'components/vapor/model/', true, true);
return true;


model/vapor/vaporvehicle.class.php
<?php
/*
 * Copyright 2012 by MODX, LLC.
 *
 * This file is part of MODX Vapor.
 *
 * Vapor is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Vapor is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Vapor; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

/**
 * A custom xPDOVehicle implementation for generic MYSQL tables and data.
 */
class vaporVehicle extends xPDOVehicle {
    public $class = 'vaporVehicle';

    /**
     * Put a representation of a MYSQL table and it's data into this vehicle.
     *
     * @param xPDOTransport $transport The transport package hosting the vehicle.
     * @param mixed &$object A reference to the artifact this vehicle will represent.
     * @param array $attributes Additional attributes represented in the vehicle.
     */
    public function put(& $transport, & $object, $attributes = array ()) {
        if (!isset ($this->payload['class'])) {
            $this->payload['class'] = $this->class;
        }
        if (is_array($object) && isset($object['table']) && isset($object['tableName'])) {
            $this->payload['object'] = $object;
        }
        parent :: put($transport, $object, $attributes);
    }

    /**
     * Install the vehicle artifact into a transport host.
     *
     * @param xPDOTransport &$transport A reference to the transport.
     * @param array $options An array of options for altering the installation of the artifact.
     * @return boolean True if the installation of the vehicle artifact was successful.
     */
    public function install(& $transport, $options) {
        $installed = false;
        $vOptions = $this->get($transport, $options);
        if (isset($vOptions['object']) && isset($vOptions['object']['tableName']) && isset($vOptions['object']['table']) && isset($vOptions['object']['data'])) {
            $tableName = $vOptions['object']['tableName'];
            /* attempt to execute the drop table if exists script */
            $dropTableQuery = isset($vOptions['object']['drop']) && !empty($vOptions['object']['drop'])
                ? $vOptions['object']['drop']
                : "DROP TABLE IF EXISTS {$transport->xpdo->escape($transport->xpdo->getOption('table_prefix', $options, '') . $tableName)}";
            $tableDropped = $transport->xpdo->exec($dropTableQuery);
            if ($tableDropped === false) {
                $transport->xpdo->log(xPDO::LOG_LEVEL_WARN, "Error executing drop table script for table {$tableName}:\n{$dropTableQuery}");
            }
            /* attempt to execute the table creation script */
            $tableCreationQuery = str_replace("[[++table_prefix]]", $transport->xpdo->getOption('table_prefix', $options, ''), $vOptions['object']['table']);
            $tableCreated = $transport->xpdo->exec($tableCreationQuery);
            if ($tableCreated !== false) {
                /* insert data rows into the table */
                if (is_array($vOptions['object']['data'])) {
                    $rowsCreated = 0;
                    foreach ($vOptions['object']['data'] as $idx => $row) {
                        $insertResult = $transport->xpdo->exec(str_replace("[[++table_prefix]]", $transport->xpdo->getOption('table_prefix', $options, ''), $row));
                        if ($insertResult === false) {
                            $transport->xpdo->log(xPDO::LOG_LEVEL_INFO, "Error inserting row {$idx} into table {$tableName}: " . print_r($transport->xpdo->errorInfo(), true));
                        } else {
                            $rowsCreated++;
                        }
                    }
                    $transport->xpdo->log(xPDO::LOG_LEVEL_INFO, "Inserted {$rowsCreated} rows into table {$tableName}");
                }
            } else {
                $transport->xpdo->log(xPDO::LOG_LEVEL_ERROR, "Could not create table {$tableName}: " . print_r($transport->xpdo->errorInfo(), true));
            }
        }
        return $installed;
    }

    /**
     * This vehicle implementation does not support uninstall.
     *
     * @param xPDOTransport &$transport A reference to the transport.
     * @param array $options An array of options for altering the uninstallation of the artifact.
     * @return boolean True, always.
     */
    public function uninstall(& $transport, $options) {
        return true;
    }
}

scripts/resolve.extension_packages.php
<?php
/*
 * Copyright 2012 by MODX, LLC.
 *
 * This file is part of MODX Vapor.
 *
 * Vapor is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Vapor is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Vapor; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

/**
 * @var xPDOTransport $transport
 * @var modSystemSetting $object
 * @var array $options
 */
if ($object instanceof modSystemSetting && $object->get('key') === 'extension_packages') {
    $extPackages = $object->get('value');
    $extPackages = $transport->xpdo->fromJSON($extPackages);
    if (!is_array($extPackages)) $extPackages = array();
    if (is_array($options) && array_key_exists('extension_packages', $options)) {
        $optPackages = $transport->xpdo->fromJSON($options['extension_packages']);
        if (is_array($optPackages)) {
            $extPackages = array_merge($extPackages, $optPackages);
        }
    }
    if (!empty($extPackages)) {
        foreach ($extPackages as $extPackage) {
            if (!is_array($extPackage)) continue;

            foreach ($extPackage as $packageName => $package) {
                if (!empty($package) && !empty($package['path'])) {
                    $package['tablePrefix'] = !empty($package['tablePrefix']) ? $package['tablePrefix'] : null;
                    $package['path'] = str_replace(array(
                        '[[++core_path]]',
                        '[[++base_path]]',
                        '[[++assets_path]]',
                        '[[++manager_path]]',
                    ),array(
                        $transport->xpdo->config['core_path'],
                        $transport->xpdo->config['base_path'],
                        $transport->xpdo->config['assets_path'],
                        $transport->xpdo->config['manager_path'],
                    ),$package['path']);
                    $transport->xpdo->addPackage($packageName,$package['path'],$package['tablePrefix']);
                    if (!empty($package['serviceName']) && !empty($package['serviceClass'])) {
                        $packagePath = str_replace('//','/',$package['path'].$packageName.'/');
                        $transport->xpdo->getService($package['serviceName'],$package['serviceClass'],$packagePath);
                    }
                }
            }
        }
    }
}